README
NAME
    POE::Component::Client::NNTP::Tail - Sends events for new articles
    posted to an NNTP newsgroup

VERSION
    This documentation describes version 0.01.

SYNOPSIS
       use POE qw( Component::Client::NNTP::Tail );
       use Email::Simple;
     
   POE::Component::Client::NNTP::Tail->spawn(
         NNTPServer  => 'nntp.perl.org',
         Group       => 'perl.cpan.testers',
       );
     
   POE::Session->create(
         package_states => [
           main => [qw(_start new_header got_article)]
         ],
       );
     
   POE::Kernel->run;
     
   # register for NNTP tail events
       sub _start {
         $_[KERNEL]->post( 'perl.cpan.testers' => 'register' );
         return;
       }
     
   # get articles with subject 'FAIL' as 'got_article' events
       sub new_header {
         my ($article_id, $lines) = @_[ARG0, ARG1];
         my $article = Email::Simple->new( join("\r\n", @$lines) );
         if ( $article->header('Subject') =~ /^FAIL/ ) {
           $_[KERNEL]->post( 
             'perl.cpan.testers' => 'get_article' => $article_id 
           );
         }
         return;
       }
     
   # find and print perl version components to terminal
       sub got_article {
         my ($article_id, $lines) = @_[ARG0, ARG1];
         for my $text ( reverse @$lines ) {
           if ( $text =~ /^Summary of my perl5 \(([^)]+)\)/ ) {
             print "$1\n";
             last;
           }
         }
         return;
       }

DESCRIPTION
    This component periodically polls an NNTP newsgroup and posts POE events
    to component listeners as new articles are available. These events
    contains the article ID and header text for the given articles. This
    component also facilitates retrieving the full text of a particular
    article of interest.

    Internally, it uses POE::Component::Client::NNTP to manage the NNTP
    session.

USAGE
    Spawn a new component session for each newsgroup to follow. Send the
    "register" event to specify an event to sent back when new articles
    arrive. Handle the new article event. Optionally, send the "get_article"
    event to request the full text of the article.

  spawn
       POE::Component::Client::NNTP::Tail->spawn(
         NNTPServer  => 'nntp.perl.org',
         Group       => 'perl.cpan.testers',
       );

    The "spawn" class method launches a new POE::Session to follow a given
    newsgroup. The "NNTPServer" and "Group" arguments are required, all
    other arguments are optional:

    *   NNTPServer (required) -- name or IP address of the NNTP server

    *   Group (required) -- newsgroup to follow

    *   Interval -- minimum number of seconds between checks for new
        messages (defaults to 60)

    *   Alias -- POE::Session alias name (defaults to the newsgroup name)

    *   Port -- server port for NNTP connections

    *   LocalAddr -- local address for outbound IP connection

    *   Debug -- if true, a trace of events and arguments will be printed to
        STDERR

    You must spawn multiple times to follow multiple newsgroups.

INPUT EVENTS
    The component will respond to the following events.

  register
       $_[KERNEL]->post( 'perl.cpan.testers' => register => $event_name );

    This event notifies the component to post a "new_header" event to the
    sender when new articles arrive. The event will be sent using the
    $event_name provided, or will default to 'new_header'. Multiple sessions
    may register with a single POE::Component::Client::NNTP::Tail session.

  unregister
       $_[KERNEL]->post( 'perl.cpan.testers' => unregister );

    This event will stop the component from posting new_header events to the
    sender.

  get_article
       $_[KERNEL]->post( 
         'perl.cpan.testers' => get_article => $article_id => $event_name
       );

    This event requests that the full text of $article_id be returned in a
    "got_article" event. The event will be sent using the $event_name
    provided, or will default to 'got_article'.

  shutdown
       $_[KERNEL]->post( 'perl.cpan.testers' => 'shutdown' );

    This event requests that the component stop broadcasting events,
    disconnect from the NNTP server and generally stop processing.

OUTPUT EVENTS
    The component sends the following events types, though the actual event
    name may be different depending on what is specified in the "register"
    and "get_article" events.

  new_header
       ($article_id, $lines) = @_[ARG0, ARG1];

    The "new_header" event is sent when new articles are found in the
    newsgroup. The $lines argument is a reference to an array of lines that
    contain the article header. Lines have had newlines removed.

  got_article
       ($article_id, $lines) = @_[ARG0, ARG1];

    The "got_article" event is sent when the full text of an article is
    retrieved. The $lines argument is a reference to an array of lines that
    contain the full article, including header and body text. Lines have had
    newlines removed.

BUGS
    Please report any bugs or feature using the CPAN Request Tracker. Bugs
    can be submitted through the web interface at
    <http://rt.cpan.org/Dist/Display.html?Queue=POE-Component-Client-NNTP-Ta
    il>

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

SEE ALSO
    *   POE

    *   POE::Component::Client::NNTP

AUTHOR
    David A. Golden (DAGOLDEN)

    Portions based on or inspired by code in
    POE::Component::SmokeBox::Uploads::NNTP by Chris Williams.

COPYRIGHT AND LICENSE
    Copyright (c) 2008 by David A. Golden. All rights reserved.

    Licensed under Apache License, Version 2.0 (the "License"). You may not
    use this file except in compliance with the License. A copy of the
    License was distributed with this file or you may obtain a copy of the
    License from http://www.apache.org/licenses/LICENSE-2.0

    Files produced as output though the use of this software, shall not be
    considered Derivative Works, but shall be considered the original work
    of the Licensor.

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.



Changes
Revision history for POE-Component-Client-NNTP-Tail

{{$NEXT}}

0.04      2019-12-05 19:56:42+00:00 Europe/London
    - Add TimeOut option and handler for 503 errors

0.03      2011-08-19 09:38:35 America/New_York
    - Squelch some test warnings

0.02      2011-08-18 23:47:40 America/New_York
    - Fix issue where polling stops if server disappers (Chris Williams)

0.01 Tue Sep 16 12:25:19 EDT 2008
    - initial release


t/lib/DummyServer.pm
# Adapted from synopsis of PoCo::Server::NNTP
package DummyServer;

use strict;
use warnings;
use Carp::POE qw/croak/;
use Email::Simple;
use Email::Date::Format qw/email_date/;
use POE qw(Component::Server::NNTP);

# Our quick-n-dirty NNTP database :-)
# This is a HoA with keys being group names and value an array ref of msg-ids
my %Groups;
# This is a hash with msg-id keys and values containing articles
my %Messages;
# This is a hash with msg-id keys and values containing headers
my %Headers;

# A quick & dirty template to generate messages

#--------------------------------------------------------------------------#
# Methods
#--------------------------------------------------------------------------#

sub spawn {
  my ($class, %args) = @_;
  croak "port argument required" unless $args{port};
  
  my $self = bless \%args, $class; 
  $self->{session_id} = POE::Session->create(
    heap => $self,
    args => $self,
    package_states => [ 
      $class => [ qw(
        _start
        shutdown
        nntpd_listener_failed
        nntpd_connection
        nntpd_disconnected
        nntpd_cmd_post
        nntpd_cmd_ihave
        nntpd_cmd_slave
        nntpd_cmd_newnews
        nntpd_cmd_newgroups
        nntpd_cmd_list
        nntpd_cmd_group
        nntpd_cmd_article
        nntpd_cmd_head
      ) ]
    ],
    options => { trace => 0 },
  );
  return $self;
}

my $msg_count = 1;
sub add_article {
  my ($self, $group, $text) = @_;
  my $article = Email::Simple->new($text);

  # ensure we have a message id, date and subject
  my $id = "<$msg_count\@example.com>";
  $msg_count++;
  my $date = $article->header('Date') || email_date(time);
  my $from = $article->header('From') || 'anonymous@example.com';
  my $subject = $article->header('Subject') || "";

  # set required headers
  $article->header_set('Newsgroups', $group);
  $article->header_set('Path', $group);
  $article->header_set('Message-ID', $id);
  $article->header_set('Date', $date);
  $article->header_set('Subject', $subject);
  $article->header_set('From', $from);

  # store article
  my $crlf = $article->crlf;
  $Messages{$id} =  $article->as_string;
  $Headers{$id} = $article->header_obj->as_string;

  push @{$Groups{lc $group}}, $id;
  return;
}

sub _debug {
  my $where = (caller(1))[3];
  $where =~ s{.*::}{DS::};
  my @args = @_[ARG0 .. $#_];
  for ( @args ) { 
    $_ = 'undef' if not defined $_;
  }
  my $args = @args ? join( " " => "", (map { "'$_'" } @args), "" ) : ""; 
  warn "$where->($args)\n";
  return;
}

#--------------------------------------------------------------------------#
# Event handlers
#--------------------------------------------------------------------------#

sub _start {
  my ($kernel,$heap) = @_[KERNEL,HEAP];
  &_debug if $heap->{Debug};
  print "# starting DummyServer on port $heap->{port}\n";
  $heap->{nntpd} = POE::Component::Server::NNTP->spawn( 
      alias   => 'nntpd', 
      posting => 0, 
      port    => $heap->{port},
  );
  $heap->{clients} = { };
  $kernel->alias_set( 'DummyServer' );
  return;
}

sub shutdown {
  my ($kernel,$heap) = @_[KERNEL,HEAP];
  &_debug if $heap->{Debug};
  $kernel->call( $heap->{nntpd_id}, 'unregister', 'all' );
  $heap->{nntpd}->shutdown;
  delete $heap->{nntpd};
  $kernel->alias_remove( $_ ) for $kernel->alias_list( $_[SESSION] );
  return;
}

sub nntpd_listener_failed {
  my ($kernel,$heap,$op,$numeric,$string) = @_[KERNEL,HEAP,ARG0,ARG1,ARG2];
  &_debug if $heap->{Debug};
  print "# Couldn't open socket: $op : $string\n";
  return;
}

sub nntpd_connection {
  my ($kernel,$heap,$sender,$client_id) = @_[KERNEL,HEAP,SENDER,ARG0];
  &_debug if $heap->{Debug};
  $heap->{clients}->{ $client_id } = { };
  return;
}

sub nntpd_disconnected {
  my ($kernel,$heap,$sender,$client_id) = @_[KERNEL,HEAP,SENDER,ARG0];
  &_debug if $heap->{Debug};
  delete $heap->{clients}->{ $client_id };
  return;
}

sub nntpd_cmd_slave {
  my ($kernel,$heap,$sender,$client_id) = @_[KERNEL,HEAP,SENDER,ARG0];
  &_debug if $heap->{Debug};
  $kernel->post( $sender, 'send_to_client', $client_id, '202 slave status noted' );
  return;
}

sub nntpd_cmd_post {
  my ($kernel,$heap,$sender,$client_id) = @_[KERNEL,HEAP,SENDER,ARG0];
  &_debug if $heap->{Debug};
  $kernel->post( $sender, 'send_to_client', $client_id, '440 posting not allowed' );
  return;
}

sub nntpd_cmd_ihave {
  my ($kernel,$heap,$sender,$client_id) = @_[KERNEL,HEAP,SENDER,ARG0];
  &_debug if $heap->{Debug};
  $kernel->post( $sender, 'send_to_client', $client_id, '435 article not wanted' );
  return;
}

sub nntpd_cmd_newnews {
  my ($kernel,$heap,$sender,$client_id) = @_[KERNEL,HEAP,SENDER,ARG0];
  &_debug if $heap->{Debug};
  $kernel->post( $sender, 'send_to_client', $client_id, '230 list of new articles follows' );
  $kernel->post( $sender, 'send_to_client', $client_id, '.' );
  return;
}

sub nntpd_cmd_newgroups {
  my ($kernel,$heap,$sender,$client_id) = @_[KERNEL,HEAP,SENDER,ARG0];
  &_debug if $heap->{Debug};
  $kernel->post( $sender, 'send_to_client', $client_id, '231 list of new newsgroups follows' );
  $kernel->post( $sender, 'send_to_client', $client_id, '.' );
  return;
}

sub nntpd_cmd_list {
  my ($kernel,$heap,$sender,$client_id) = @_[KERNEL,HEAP,SENDER,ARG0];
  &_debug if $heap->{Debug};
  $kernel->post( $sender, 'send_to_client', $client_id, '215 list of newsgroups follows' );
  foreach my $group ( keys %Groups ) {
    my $reply = join ' ', $group, scalar @{ $Groups{$group} } + 1, 1, 'n';
    $kernel->post( $sender, 'send_to_client', $client_id, $reply );
  }
  $kernel->post( $sender, 'send_to_client', $client_id, '.' );
  return;
}

sub nntpd_cmd_group {
  my ($kernel,$heap,$sender,$client_id,$group) 
    = @_[KERNEL,HEAP,SENDER,ARG0,ARG1];
  &_debug if $heap->{Debug};
  $group = lc $group;
  unless ( $group && exists $Groups{$group} ) { 
     $kernel->post( $sender, 'send_to_client', $client_id, '411 no such news group' );
     return;
  }
  my $last = scalar @{$Groups{$group}};
  $kernel->post( $sender, 'send_to_client', $client_id, "211 $last 1 $last $group group selected" );
  $_[HEAP]->{clients}->{ $client_id } = { group => $group };
  return;
}

sub nntpd_cmd_article {
  my ($kernel,$heap,$sender,$client_id,$article) 
    = @_[KERNEL,HEAP,SENDER,ARG0,ARG1];
  &_debug if $heap->{Debug};
  my $group = $_[HEAP]->{clients}{$client_id}{group};

  my ($msg_id, $set_current) = _validate_article_id(@_, $group);

  if ($msg_id) { # article exists
    $_[HEAP]->{clients}{$client_id}{current} = $article if $set_current;

    $kernel->post( $sender, 'send_to_client', $client_id, "220 $article $msg_id article retrieved - head and body follow" );
    $kernel->post( $sender, 'send_to_client', $client_id, $Messages{$msg_id});
    $kernel->post( $sender, 'send_to_client', $client_id, '.' );
  }

  return;
}

sub nntpd_cmd_head {
  my ($kernel,$heap,$sender,$client_id,$article) 
    = @_[KERNEL,HEAP,SENDER,ARG0,ARG1];
  &_debug if $heap->{Debug};
  my $group = $_[HEAP]->{clients}{$client_id}{group};

  my ($msg_id, $set_current) = _validate_article_id(@_, $group);

  if ($msg_id) { # article exists
    $_[HEAP]->{clients}{$client_id}{current} = $article if $set_current;

    $kernel->post( $sender, 'send_to_client', $client_id, "221 $article $msg_id article retrieved - head follows" );
    $kernel->post( $sender, 'send_to_client', $client_id, $Headers{$msg_id});
    $kernel->post( $sender, 'send_to_client', $client_id, '.' );
  }

  return;
}

sub _validate_article_id {
  my ($kernel,$heap,$sender,$client_id,$article, $group) 
    = @_[KERNEL,HEAP,SENDER,ARG0,ARG1,ARG2];
  &_debug if $heap->{Debug};

  $article = 1 unless $article;
  # ARTICLE <msg_id>
  if ( $article =~ /^<.*>$/ ) {
    if ( !defined $Messages{$article} ) {
     $kernel->post( $sender, 'send_to_client', $client_id, '430 no such article found' );
     return;
    }
    else {
      return $article;
    }
  }
  # ARTICLE [nnn]
  elsif ( $article =~ /^\d+$/ ) {
    if ( !defined $group ) {
      $kernel->post( $sender, 'send_to_client', $client_id, '412 no newsgroup selected' );
      return;
    }
    else {
      my $last = scalar @{$Groups{$group}} + 1;
      if ( $article < 1 || $article > $last ) {
        $kernel->post( $sender, 'send_to_client', $client_id, '423 no such article number' );
        return;
      }
      else {
        return ($Groups{$group}[$article-1], 1);
      }
    }
  }
  # default fallthrough
  $kernel->post( $sender, 'send_to_client', $client_id, '423 no such article number' );
  return;
}

1;


Build.PL
# Copyright (c) 2008 by David Golden. All rights reserved.
# Licensed under Apache License, Version 2.0 (the "License").
# You may not use this file except in compliance with the License.
# A copy of the License was distributed with this file or you may obtain a 
# copy of the License from http://www.apache.org/licenses/LICENSE-2.0

use 5.006;
use strict;
use warnings;
use lib 'inc';
eval "require Pod::WikiDoc";
my $class = $@ ? "Module::Build" : "Module::Build::WikiDoc";
eval "require $class";
my $build = $class->new( 
    module_name => 'POE::Component::Client::NNTP::Tail',
    dist_author => 'David Golden <dagolden@cpan.org>',
    license => 'apache',
    create_readme => 1,
    create_makefile_pl  => 'traditional',
    requires => {
        'perl' => '5.006',
        'Carp::POE' => 0,
        'Params::Validate' => 0,
        'POE'  => 1.0,
        'POE::Component::Client::NNTP' => 2.10,
    },
    build_requires => {
        'Test::More' => 0.62,
        'Email::Date::Format' => 1.002,
        'Email::Simple' => 2.003,
        'POE::Component::Server::NNTP' => 1.02,
    },
    meta_add            => { 
        no_index => { 
            directory => [ qw/ examples inc t /],
        } 
    },
);
$build->create_build_script;


Makefile.PL
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.17;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Sends events for new articles posted to an NNTP newsgroup",
  "AUTHOR" => "David Golden <dagolden\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.17"
  },
  "DISTNAME" => "POE-Component-Client-NNTP-Tail",
  "LICENSE" => "apache",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "POE::Component::Client::NNTP::Tail",
  "PREREQ_PM" => {
    "Carp::POE" => 0,
    "POE" => 0,
    "POE::Component::Client::NNTP" => 0,
    "POE::Component::Server::NNTP" => 0,
    "Params::Validate" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Email::Date::Format" => 0,
    "Email::Simple" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "POE::Component::Server::NNTP" => 0,
    "Test::More" => 0,
    "lib" => 0
  },
  "VERSION" => "0.05",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp::POE" => 0,
  "Email::Date::Format" => 0,
  "Email::Simple" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "POE" => 0,
  "POE::Component::Client::NNTP" => 0,
  "POE::Component::Server::NNTP" => 0,
  "Params::Validate" => 0,
  "Test::More" => 0,
  "lib" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/POE/Component/Client/NNTP/Tail.pm
package POE::Component::Client::NNTP::Tail;

use 5.006;
use strict;
use warnings;
# ABSTRACT: Sends events for new articles posted to an NNTP newsgroup
# VERSION

our $VERSION = '0.05';

use Carp::POE;
use Params::Validate;
use POE qw(Component::Client::NNTP);

my %spawn_args = (
  # required
  Group         => 1,
  NNTPServer    => 1,
  # optional with defaults
  Interval      => { default => 60 },
  TimeOut       => { default => 30 },
  # purely optional
  Port          => 0,
  LocalAddr     => 0,
  Alias         => 0,
  Debug         => 0,
);

sub spawn {
  my $class = shift;
  my %opts = validate( @_, \%spawn_args );

  POE::Session->create(
    heap => \%opts,
    package_states => [
      # nntp component events
      $class => {
        nntp_connected    => '_nntp_connected',
        nntp_registered   => '_nntp_registered',
        nntp_socketerr    => '_nntp_socketerr',
        nntp_disconnected => '_nntp_disconnected',
        nntp_200	        => '_nntp_server_ready',
        nntp_201          => '_nntp_server_ready',
        nntp_211	        => '_nntp_group_selected',
        nntp_220	        => '_nntp_got_article',
        nntp_221	        => '_nntp_got_head',
        nntp_411          => '_nntp_no_group',
        nntp_423          => '_nntp_no_article',
        nntp_430          => '_nntp_no_article',
        nntp_503          => '_nntp_503_error',
      },
      # session events
      $class => [ qw( _start _stop _child  ) ],
      # internal events
      $class => [ qw( _poll _reconnect ) ],
      # API events
      $class => [ qw( register unregister get_article shutdown ) ],
    ],
  );
}

sub _debug {
  my $where = (caller(1))[3];
  $where =~ s{.*::}{P::C::C::N::T::};
  my @args = @_[ARG0 .. $#_];
  for ( @args ) {
    $_ = 'undef' if not defined $_;
  }
  my $args = @args ? join( " " => "", (map { "'$_'" } @args), "" ) : "";
  warn "$where->($args)\n";
  return;
}

#--------------------------------------------------------------------------#
# session events
#--------------------------------------------------------------------------#

sub _start {
  my ( $kernel, $session, $heap ) = @_[KERNEL, SESSION, HEAP];
  &_debug if $heap->{Debug};

  # alias defaults to group name if not otherwise set
  $heap->{Alias} = $heap->{Group} unless exists $heap->{Alias};
  $kernel->alias_set($heap->{Alias});

  # setup NNTP including optional args if defined;
  my %nntp_args;
  for my $k ( qw/NNTPServer Port LocalAddr TimeOut/ ) {
    $nntp_args{$k} = $heap->{$k} if exists $heap->{$k};
  }
  my $alias = "NNTP-Client-" . $session->ID;
  $heap->{nntp} = POE::Component::Client::NNTP->spawn($alias,\%nntp_args);
  $heap->{nntp_id} = $heap->{nntp}->session_id;

  # start NNTP connection
  $kernel->yield( '_reconnect' );
  return;
}

# ignore these
sub _child {
  my ( $kernel, $heap ) = @_[KERNEL, HEAP];
  &_debug if $heap->{Debug};
}

sub _stop {
  my ( $kernel, $session, $heap ) = @_[KERNEL, SESSION, HEAP];
  &_debug if $heap->{Debug};
  $kernel->call( $session, 'shutdown' ) if $heap->{nntp};
  return;
}

#--------------------------------------------------------------------------#
# events from our clients
#--------------------------------------------------------------------------#

#--------------------------------------------------------------------------#
# register -- [EVENT]
#
# EVENT - event to dispatch to the registered session on receipt of new
#         headers; defaults to "new_header"
#--------------------------------------------------------------------------#

sub register {
  my ($kernel, $heap, $sender) = @_[KERNEL, HEAP, SENDER];
  &_debug if $heap->{Debug};
  my ($event) = $_[ARG0] || 'new_header';
  $kernel->refcount_increment( $sender->ID, __PACKAGE__ );
  $heap->{listeners}{$sender->ID} = $event;
  return;
}

#--------------------------------------------------------------------------#
# unregister --
#
# removes sender registration
#--------------------------------------------------------------------------#

sub unregister {
  my ($kernel, $heap, $sender) = @_[KERNEL, HEAP, SENDER];
  &_debug if $heap->{Debug};
  $kernel->refcount_decrement( $sender->ID, __PACKAGE__ );
  delete $heap->{listeners}{$sender->ID};
  return;
}

#--------------------------------------------------------------------------#
# get_article -- ARTICLE_ID, EVENT
#
# request ARTICLE_ID be retrieved and returned via EVENT or 'got_article
# if not specified
#--------------------------------------------------------------------------#

sub get_article {
  my ($kernel, $heap, $sender) = @_[KERNEL, HEAP, SENDER];
  &_debug if $heap->{Debug};
  my ($article_id, $return_event) = @_[ARG0, ARG1];
  $return_event ||= 'got_article';
  # store requesting session and desired return event
  push @{$heap->{requests}{$article_id}}, [$sender, $return_event];
  $kernel->post( $heap->{nntp_id} => article => $article_id );
  return;
}

#--------------------------------------------------------------------------#
# shudown
#--------------------------------------------------------------------------#

sub shutdown {
  my ( $kernel, $heap ) = @_[KERNEL, HEAP];
  &_debug if $heap->{Debug};
  # unregister anyone that didn't do it themselves
  for my $listener ( keys %{ $heap->{listeners} } ) {
    $kernel->refcount_decrement( $listener, __PACKAGE__ );
    delete $heap->{listeners}{$listener};
  }
  $kernel->alarm_remove_all();
  $kernel->call( $heap->{nntp_id} => 'unregister' => 'all' );
  $kernel->call( $heap->{nntp_id} => 'shutdown' );
  delete $heap->{nntp};
  $kernel->alias_remove($_) for $kernel->alias_list;
  return;
}

#--------------------------------------------------------------------------#
# our internal events
#--------------------------------------------------------------------------#

# if connected, check for new messages, otherwise reconnect
sub _poll {
  my ( $kernel, $heap ) = @_[KERNEL, HEAP];
  &_debug if $heap->{Debug};
  if ( $heap->{connected} ) {
    $kernel->post( $heap->{nntp_id} => group => $heap->{Group} );
  }
  else {
    $kernel->yield( '_reconnect' );
  }
  $kernel->delay( '_poll' => $heap->{Interval} );
  return;
}

# connect to NNTP server
sub _reconnect {
  my ( $kernel, $heap ) = @_[KERNEL, HEAP];
  &_debug if $heap->{Debug};
  $kernel->post( $heap->{nntp_id} => 'connect' );
  return;
}

#--------------------------------------------------------------------------#
# events from NNTP client
#--------------------------------------------------------------------------#

# ignore event
sub _nntp_registered {
  my ($kernel, $heap, $sender) = @_[KERNEL, HEAP, SENDER];
  &_debug if $heap->{Debug};
  return;
}

# ignore event
sub _nntp_connected {
  my ($kernel, $heap, $sender) = @_[KERNEL, HEAP, SENDER];
  &_debug if $heap->{Debug};
  return;
}

# if connection can't be made, wait for next poll period to try again
sub _nntp_socketerr {
  my ($kernel, $heap, $sender) = @_[KERNEL, HEAP, SENDER];
  &_debug if $heap->{Debug};
  my ($error) = $_[ARG0];
  warn "Socket error: $error\n";
  $heap->{connected} = 0;
  $kernel->delay( '_reconnect' => $heap->{Interval} );
  return;
}

# if we time-out, just note it and wait for next poll to reconnect
sub _nntp_disconnected {
  my ($kernel, $heap, $sender) = @_[KERNEL, HEAP, SENDER];
  &_debug if $heap->{Debug};
  $heap->{connected} = 0;
  return;
}

# once connected, start polling loop
sub _nntp_server_ready {
  my ($kernel, $heap, $sender) = @_[KERNEL, HEAP, SENDER];
  &_debug if $heap->{Debug};
  $heap->{connected} = 1;
  $kernel->yield( '_poll' );
  undef;
}

# if the group doesn't exist, then we shut ourselves down
sub _nntp_no_group {
  my ($kernel, $heap, $sender) = @_[KERNEL, HEAP, SENDER];
  &_debug if $heap->{Debug};
  warn "No such newsgroup $heap->{Group}";
  $kernel->yield( 'shutdown' );
  return;
}

# if the article doesn't exist, warn about it
sub _nntp_no_article {
  my ($kernel, $heap, $sender) = @_[KERNEL, HEAP, SENDER];
  &_debug if $heap->{Debug};
  warn "Couldnt find article in $heap->{Group}\n";
  return;
}

# 503 error in response to group query reconnect
sub _nntp_503_error {
  my ($kernel, $heap, $sender) = @_[KERNEL, HEAP, SENDER];
  &_debug if $heap->{Debug};
  $heap->{connected} = 0;
  $kernel->delay( '_reconnect' => $heap->{Interval} );
  return;
}

# if there are new articles, request their headers
# also schedules the next check
sub _nntp_group_selected {
  my ($kernel, $heap, $sender) = @_[KERNEL, HEAP, SENDER];
  &_debug if $heap->{Debug};
  my ($estimate,$first,$last,$group) = split( /\s+/, $_[ARG0] );

  # first time, we won't know last_article, so skip to the end
  if ( exists $heap->{last_article} ) {
    # fetch new headers or articles only if people are listening
    for my $article_id ( $heap->{last_article} + 1 .. $last ) {
      if ( scalar keys %{ $heap->{listeners} } ) {
        $kernel->post( $sender => head => $article_id );
      }
    }
  }
  $heap->{last_article} = $last;
  $kernel->delay( '_poll' => $heap->{Interval} );
  return;
}

# notify listeners of new header
sub _nntp_got_head {
  my ($kernel, $heap, $sender) = @_[KERNEL, HEAP, SENDER];
  &_debug if $heap->{Debug};
  my ($response, $lines) = @_[ARG0, ARG1];
  my ($article_id) = split " ", $response;
  for my $who ( keys %{ $heap->{listeners} } ) {
    $kernel->post( $who => $heap->{listeners}{$who} => $article_id, $lines );
  }
  return;
}

# return article to request queue
sub _nntp_got_article {
  my ($kernel, $heap, $sender) = @_[KERNEL, HEAP, SENDER];
  &_debug if $heap->{Debug};
  my ($response, $lines) = @_[ARG0, ARG1];
  my ($article_id) = split " ", $response;
  # dispatch for all entries in the request queue for this article
  for my $request ( @{$heap->{requests}{$article_id}} ) {
    my ($who, $event) = @$request;
    $kernel->post( $who, $event, $article_id, $lines );
  }
  # clear the request queue
  delete $heap->{requests}{$article_id};
  return;
}

1;

__END__

=pod

=begin wikidoc

= SYNOPSIS

  use POE qw( Component::Client::NNTP::Tail );
  use Email::Simple;

  POE::Component::Client::NNTP::Tail->spawn(
    NNTPServer  => 'nntp.perl.org',
    Group       => 'perl.cpan.testers',
  );

  POE::Session->create(
    package_states => [
      main => [qw(_start new_header got_article)]
    ],
  );

  POE::Kernel->run;

  # register for NNTP tail events
  sub _start {
    $_[KERNEL]->post( 'perl.cpan.testers' => 'register' );
    return;
  }

  # get articles with subject 'FAIL' as 'got_article' events
  sub new_header {
    my ($article_id, $lines) = @_[ARG0, ARG1];
    my $article = Email::Simple->new( join("\r\n", @$lines) );
    if ( $article->header('Subject') =~ /^FAIL/ ) {
      $_[KERNEL]->post(
        'perl.cpan.testers' => 'get_article' => $article_id
      );
    }
    return;
  }

  # find and print perl version components to terminal
  sub got_article {
    my ($article_id, $lines) = @_[ARG0, ARG1];
    for my $text ( reverse @$lines ) {
      if ( $text =~ /^Summary of my perl5 \(([^)]+)\)/ ) {
        print "$1\n";
        last;
      }
    }
    return;
  }

= DESCRIPTION

This component periodically polls an NNTP newsgroup and posts POE events to
component listeners as new articles are available.  These events contains the
article ID and header text for the given articles.  This component also
facilitates retrieving the full text of a particular article of interest.

Internally, it uses [POE::Component::Client::NNTP] to manage the NNTP session.

= USAGE

Spawn a new component session for each newsgroup to follow. Send the
{register} event to specify an event to sent back when new articles arrive.
Handle the new article event. Optionally, send the {get_article} event to
request the full text of the article.

== spawn

  POE::Component::Client::NNTP::Tail->spawn(
    NNTPServer  => 'nntp.perl.org',
    Group       => 'perl.cpan.testers',
  );

The {spawn} class method launches a new POE::Session to follow a given
newsgroup.  The {NNTPServer} and {Group} arguments are required, all other
arguments are optional:

* NNTPServer (required) -- name or IP address of the NNTP server
* Group (required) -- newsgroup to follow
* Interval -- minimum number of seconds between checks for new messages
(defaults to 60)
* TimeOut -- number of seconds to wait for a response from server
(defaults to 30)
* Alias -- POE::Session alias name (defaults to the newsgroup name)
* Port -- server port for NNTP connections
* LocalAddr -- local address for outbound IP connection
* Debug -- if true, a trace of events and arguments will be printed to STDERR

You must spawn multiple times to follow multiple newsgroups.

= INPUT EVENTS

The component will respond to the following events.

== register

  $_[KERNEL]->post( 'perl.cpan.testers' => register => $event_name );

This event notifies the component to post a {new_header} event to the sender
when new articles arrive.  The event will be sent using the {$event_name}
provided, or will default to 'new_header'.  Multiple sessions may register
with a single POE::Component::Client::NNTP::Tail session.

== unregister

  $_[KERNEL]->post( 'perl.cpan.testers' => unregister );

This event will stop the component from posting new_header events to the
sender.

== get_article

  $_[KERNEL]->post(
    'perl.cpan.testers' => get_article => $article_id => $event_name
  );

This event requests that the full text of {$article_id} be returned in a
{got_article} event.  The event will be sent using the {$event_name}
provided, or will default to 'got_article'.

== shutdown

  $_[KERNEL]->post( 'perl.cpan.testers' => 'shutdown' );

This event requests that the component stop broadcasting events,
disconnect from the NNTP server and generally stop processing.

= OUTPUT EVENTS

The component sends the following events types, though the actual event
name may be different depending on what is specified in the {register} and
{get_article} events.

== new_header

  ($article_id, $lines) = @_[ARG0, ARG1];

The {new_header} event is sent when new articles are found in the newsgroup.
The {$lines} argument is a reference to an array of lines that contain the
article header.  Lines have had newlines removed.

== got_article

  ($article_id, $lines) = @_[ARG0, ARG1];

The {got_article} event is sent when the full text of an article is retrieved.
The {$lines} argument is a reference to an array of lines that contain the full
article, including header and body text.  Lines have had newlines removed.

= BUGS

Please report any bugs or feature using the CPAN Request Tracker.
Bugs can be submitted through the web interface at
[http://rt.cpan.org/Dist/Display.html?Queue=POE-Component-Client-NNTP-Tail]

When submitting a bug or request, please include a test-file or a patch to an
existing test-file that illustrates the bug or desired feature.

= SEE ALSO

* [POE]
* [POE::Component::Client::NNTP]

=end wikidoc

=cut

